File: rust/src/2402.meeting-rooms-iii/solution.rs
// Created by none at 2025/07/11 09:57
// leetgo: dev
// https://leetcode.com/problems/meeting-rooms-iii/

#![allow(unused_imports)]

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::cmp::Reverse;
use std::collections::*;
use std::mem::swap;

impl Solution {
    pub fn most_booked(n: i32, meetings: Vec<Vec<i32>>) -> i32 {
        let n = n as usize;
        let mut cnt = vec![0; n];
        let mut meetings = meetings;
        meetings.sort_unstable();
        // println!("{:?}", meetings);

        let mut free_room = BinaryHeap::from_iter((0..n).map(|i| Reverse(i)));
        let mut used_room: BinaryHeap<Reverse<(i64, usize)>> = BinaryHeap::new();

        for m in &meetings {
            while let Some(Reverse((end, i))) = used_room.peek() {
                if *end <= m[0] as i64 {
                    free_room.push(Reverse(*i));
                    used_room.pop();
                } else {
                    break;
                }
            }
            let i;
            let end;
            if free_room.is_empty() {
                // skip to first time can release room
                let room = used_room.pop().unwrap().0;
                end = room.0 + (m[1] - m[0]) as i64;
                i = room.1;
            } else {
                i = free_room.pop().unwrap().0;
                end = m[1] as i64;
            }
            cnt[i] += 1;
            used_room.push(Reverse((end, i)));
        }

        let mut i = 0;
        for j in 1..n {
            if cnt[j] > cnt[i] {
                i = j
            }
        }
        // println!("{cnt:?} {}", cnt[i]);
        // assert_eq!(cnt.iter().sum::<i32>(), meetings.len() as i32);

        i as i32
    }
}

// @lc code=end

fn main() -> Result<()> {
    let n: i32 = deserialize(&read_line()?)?;
    let meetings: Vec<Vec<i32>> = deserialize(&read_line()?)?;
    let ans: i32 = Solution::most_booked(n, meetings).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}

File: go/2402.meeting-rooms-iii/solution.go
// Created by none at 2025/12/27 12:22
// leetgo: dev
// https://leetcode.com/problems/meeting-rooms-iii/

package main

import (
	"bufio"
	"cmp"
	"fmt"
	"os"
	"slices"

	"github.com/emirpasic/gods/v2/queues/priorityqueue"
	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

type Pair struct{ end, index int }

func pairComparator(a, b Pair) int {
	if a.end != b.end {
		return cmp.Compare(a.end, b.end) // smaller end comes first
	}
	return cmp.Compare(a.index, b.index) // tie-break by smaller index
}
func mostBooked(n int, meetings [][]int) int {
	slices.SortFunc(meetings, func(a, b []int) int { return a[0] - b[0] })

	unused := priorityqueue.New[int]()
	for i := range n {
		unused.Enqueue(i)
	}
	// fmt.Println(meetings)
	using := priorityqueue.NewWith(pairComparator)
	cnt := make([]int, n)
	for _, m := range meetings {
		start, end := m[0], m[1]
		for !using.Empty() {
			p, _ := using.Peek()
			if p.end <= start {
				using.Dequeue()
				unused.Enqueue(p.index)
			} else {
				break
			}
		}
		if !unused.Empty() {
			i, _ := unused.Dequeue()
			cnt[i]++
			using.Enqueue(Pair{end, i})
		} else {
			p, _ := using.Dequeue()
			cnt[p.index]++
			using.Enqueue(Pair{p.end + (end - start), p.index})
		}
	}

	ans := 0
	for i, x := range cnt {
		if x > cnt[ans] {
			ans = i
		}
	}
	return ans
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	n := Deserialize[int](ReadLine(stdin))
	meetings := Deserialize[[][]int](ReadLine(stdin))
	ans := mostBooked(n, meetings)

	fmt.Println("\noutput:", Serialize(ans))
}
